resharper/src/resharper-unity/CSharp/Daemon/Stages/Analysis/DrawGizmoAttributeProblemAnalyzer.cs

```diff
@@ -0,0 +1,89 @@
+            IHighlightingConsumer consumer)
+        {
+            if (!(element.TypeReference?.Resolve().DeclaredElement is ITypeElement attributeTypeElement))
+                return;
+
+            // Otherwise we'll treat it as targeting a method
+            if (element.Target == AttributeTarget.Return)
+                return;
+
+            if (attributeTypeElement.GetClrName().Equals(KnownTypes.DrawGizmo))
+            {
+                var methodDeclaration = MethodDeclarationNavigator.GetByAttribute(element);
+                CheckMethodDeclaration(methodDeclaration, element, consumer);
+            }
+        }
+
+        private void CheckMethodDeclaration(IMethodDeclaration methodDeclaration, IAttribute element, IHighlightingConsumer consumer)
+        {
+            if (methodDeclaration == null)
+                return;
+            
+            var predefinedType = myPredefinedTypeCache.GetOrCreatePredefinedType(element.GetPsiModule());
+            var gizmoType = TypeFactory.CreateTypeByCLRName("UnityEditor.GizmoType", predefinedType.Module);
+            var componentType = TypeFactory.CreateTypeByCLRName("UnityEngine.Component", predefinedType.Module);
+          
+
+            var expectedDeclaration = new MethodSignature(predefinedType.Void, true,
+                new[] {componentType, gizmoType},
+                new[] {"component", "gizmoType"});
+
+            var match = expectedDeclaration.Match(methodDeclaration);
+            
+            if (methodDeclaration.Params.ParameterDeclarations.Count == 2)
+            {
+                var parameters = methodDeclaration.Params.ParameterDeclarations;
+
+                Log.Root.Log(LoggingLevel.INFO, string.Format("component {0}:{1}:{2}", parameters[0].Type.GetTypeElement(), componentType.GetTypeElement(), parameters[0].Type.GetTypeElement()
+                    ?.IsDescendantOf(componentType.GetTypeElement())));
+                Log.Root.Log(LoggingLevel.INFO, string.Format("gizmo {0}:{1}:{2}", parameters[1].Type.GetTypeElement(), gizmoType.GetTypeElement(), parameters[1].Type.GetTypeElement()?.Equals(gizmoType?.GetTypeElement())));
+
+                if (parameters[0].Type.GetTypeElement()
+                        ?.IsDescendantOf(componentType.GetTypeElement()) == true)
+                {
+                    Log.Root.Log(LoggingLevel.INFO, "First matched");
+                    if (parameters[1].Type.GetTypeElement()?.Equals(gizmoType?.GetTypeElement()) == true)
+                    {
+                        Log.Root.Log(LoggingLevel.INFO, "second matched");
+                        match &= ~MethodSignatureMatch.IncorrectParameters;
+                    }
+                
+                    expectedDeclaration = new MethodSignature(predefinedType.Void, true,
+                        new[] {parameters[0].Type, gizmoType},
+                        new[] {parameters[0].DeclaredName, parameters[1].DeclaredName});
```

Comment:
I really like this. Means that the quick fix if the parameters are wrong will have the correct type for the first parameter. But it also means that I want it for more scenarios, too :)

Firstly, if there are one or more parameters, find the first one that derives from `Component` or `MonoBehaviour` and use this type to update the `MethodSignature`. So if someone has `Foo(MyComponent component, int whatever, string yeah, float really, double wrong)`, the inspections and quick fix will show "Change parameters to '(MyComponent, GizmoType)'" instead of losing information with "Change to '(Component, GizmoType)'". And if they do `Foo(GizmoType, MyComponent)`, or `Foo(MyComponent, int)' we'll show them the correct type for `Component` too.

Secondly, check that there are two parameters, and that they are the correct types (probably worth setting a flag from the first check for the component type).

And add test cases:
* Too many parameters with the component type in the first position
* Too many parameters with the component type NOT in the first position
* Correct first parameter and incorrect second parameter